File: python-algorithm/ds_9_bit_operation/L338-m.py
# -*- coding: utf-8 -*-



# 动态规划
class Solution:
    def countBits(self, num: int) -> List[int]:
        res = [0 for i in range(num+1)]
        for i in range(1, num+1):
            res[i] = res[i>>1] + (i&1);
        return res

File: algo_05_dynamic_plan/dp_4_special_state_transition/L338-m.cpp


//位运算+动态规划
//转移方程：状态转移函数
//P(x) = P(x/2) + (x mod 2)
//思想：一个整数和一个整数的一半（向下取余）的二进制表示只有最后一位不同，而最后一位是奇数偶数的判定

#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> res(num+1, 0);
        for(int i=1; i<=num; ++i)
            res[i] = res[i>>1] + (i&1);
        return res;
    }
};

//另外一个思路，也是动态规划 转移状态方程：dp[i] = dp[i&(i-1)] + 1;